File: backend/app/models/daily_report.py
from datetime import UTC, date, datetime
from uuid import UUID, uuid4

from sqlalchemy import Date, DateTime, ForeignKey, Integer, String, Text, UniqueConstraint
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.core.database import Base


class DailyReport(Base):
    __tablename__ = "daily_reports"
    __table_args__ = (
        UniqueConstraint(
            "department_id",
            "report_date",
            name="uq_daily_report_department_date",
        ),
    )

    id: Mapped[UUID] = mapped_column(primary_key=True, default=uuid4)
    department_id: Mapped[UUID] = mapped_column(
        ForeignKey("departments.id", ondelete="CASCADE"),
        index=True,
        nullable=False,
    )
    report_date: Mapped[date] = mapped_column(Date, index=True, nullable=False)
    status: Mapped[str] = mapped_column(String(20), default="READY", index=True)
    file_path: Mapped[str | None] = mapped_column(Text, nullable=True)
    file_size: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    created_count: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    completed_count: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    pending_count: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    in_progress_count: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    error: Mapped[str | None] = mapped_column(Text, nullable=True)
    generated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=lambda: datetime.now(UTC),
        index=True,
        nullable=False,
    )

    department: Mapped["Department"] = relationship(lazy="joined")

File: mobile/lib/models/daily_report.dart
import 'department.dart';

class DailyReportItem {
  const DailyReportItem({
    required this.id,
    required this.department,
    required this.reportDate,
    required this.status,
    required this.fileSize,
    required this.createdCount,
    required this.completedCount,
    required this.pendingCount,
    required this.inProgressCount,
    required this.generatedAt,
    this.error,
  });

  final String id;
  final DepartmentSummary department;
  final DateTime reportDate;
  final String status;
  final int fileSize;
  final int createdCount;
  final int completedCount;
  final int pendingCount;
  final int inProgressCount;
  final String? error;
  final DateTime generatedAt;

  factory DailyReportItem.fromJson(Map<String, dynamic> json) {
    return DailyReportItem(
      id: json['id'].toString(),
      department: DepartmentSummary.fromJson(
        Map<String, dynamic>.from(json['department'] as Map),
      ),
      reportDate: DateTime.parse(json['report_date'].toString()),
      status: json['status'].toString(),
      fileSize: (json['file_size'] as num?)?.toInt() ?? 0,
      createdCount: (json['created_count'] as num?)?.toInt() ?? 0,
      completedCount: (json['completed_count'] as num?)?.toInt() ?? 0,
      pendingCount: (json['pending_count'] as num?)?.toInt() ?? 0,
      inProgressCount: (json['in_progress_count'] as num?)?.toInt() ?? 0,
      error: json['error'] as String?,
      generatedAt: DateTime.parse(json['generated_at'].toString()),
    );
  }
}
